File: packages/ik_rules/lib/src/combat/boss_phase.dart
import 'dart:math' as math;

import 'package:ik_content/ik_content.dart';

import '../achievements/progress.dart';
import '../activity/xp.dart';
import '../skills/skill_actions.dart' show fishingSkillId;
import '../save/generated/save_models.dart';
import 'boss.dart';
import 'engine.dart';

class SquidlingVictoryResult {
  const SquidlingVictoryResult({
    required this.save,
    required this.xpGained,
    required this.message,
    required this.bossResumed,
  });

  final PlayerSave save;
  final num xpGained;
  final String message;
  final bool bossResumed;
}

PlayerSave beginBossAddsEncounter(
  GameDatabase db,
  PlayerSave save,
  EnemyRow bossEnemy,
  BossProfile profile,
  num pendingHp,
  String roundEndIso,
) {
  final squidling = getEnemy(db, profile.squidlingEnemyId!);
  if (squidling == null) return save;
  return save.copyWith(
    combatBossPendingId: bossEnemy.raw['Enemy ID'] as String?,
    combatBossPendingHp: pendingHp,
    combatBossAddsRemaining: profile.squidlingCount,
    combatBossAddsTriggered: true,
    combatBossInkActive: false,
    combatEnemyId: squidling.raw['Enemy ID'] as String?,
    combatEnemyHp: squidling.maximumHp,
    combatRoundStartedAt: roundEndIso,
    combatSkipEnemyAttack: false,
    combatBossSleepRoundsRemaining: null,
  );
}

SquidlingVictoryResult applySquidlingVictory(
  GameDatabase db,
  PlayerSave save,
  EnemyRow squidling,
  String roundEndIso,
) {
  final xpAmount = squidling.combatXp ?? 0;
  var next = applyXp(save, db, fishingSkillId, xpAmount).save;
  final kills = (next.statistics.values['monsters_killed'] ?? 0) + 1;
  next = next.copyWith(
    statistics: PlayerStatistics(
      values: <String, num>{...next.statistics.values, 'monsters_killed': kills},
    ),
  );
  next = recordEnemyKill(db, next, squidling.raw['Enemy ID'] as String);

  final remaining = math.max(0, (next.combatBossAddsRemaining ?? 1) - 1);
  final bossId = next.combatBossPendingId;
  final pendingHp = next.combatBossPendingHp ?? 0;

  if (remaining > 0 && bossId != null) {
    final profile = bossProfile(getEnemy(db, bossId));
    final nextSquidling = profile?.squidlingEnemyId != null
        ? getEnemy(db, profile!.squidlingEnemyId!)
        : null;
    if (nextSquidling == null) {
      return SquidlingVictoryResult(
        save: next,
        xpGained: xpAmount,
        message: 'Squidling defeated.',
        bossResumed: false,
      );
    }
    final total = profile?.squidlingCount ?? remaining + 1;
    final defeated = total - remaining;
    return SquidlingVictoryResult(
      save: next.copyWith(
        combatBossAddsRemaining: remaining,
        combatEnemyId: nextSquidling.raw['Enemy ID'] as String?,
        combatEnemyHp: nextSquidling.maximumHp,
        combatRoundStartedAt: roundEndIso,
        combatBossInkActive: false,
        combatSkipEnemyAttack: false,
      ),
      xpGained: xpAmount,
      message: 'Squidling defeated ($defeated/$total). Another emerges!',
      bossResumed: false,
    );
  }

  if (bossId == null) {
    return SquidlingVictoryResult(
      save: next,
      xpGained: xpAmount,
      message: 'Squidling defeated.',
      bossResumed: false,
    );
  }

  final boss = getEnemy(db, bossId);
  if (boss == null) {
    return SquidlingVictoryResult(
      save: next,
      xpGained: xpAmount,
      message: 'Squidling defeated.',
      bossResumed: false,
    );
  }

  return SquidlingVictoryResult(
    save: next.copyWith(
      combatBossAddsRemaining: null,
      combatBossPendingId: null,
      combatBossPendingHp: null,
      combatEnemyId: bossId,
      combatEnemyHp: pendingHp,
      combatRoundStartedAt: roundEndIso,
      combatBossInkActive: false,
      combatSkipEnemyAttack: false,
      combatBossSleepRoundsRemaining: null,
    ),
    xpGained: xpAmount,
    message: 'Squidling defeated. ${boss.displayName} returns at $pendingHp HP!',
    bossResumed: true,
  );
}

bool isSquidlingVictory(PlayerSave save, EnemyRow enemy) {
  if (!isBossAddFight(save)) return false;
  return save.combatEnemyId == enemy.raw['Enemy ID'];
}

File: src/game/combat/bossPhase.ts
import { applyXp } from '../activity/xp'
import { FISHING_SKILL_ID } from '../skills/skillActions'
import type { GameDatabase } from '../data/types'
import type { EnemyRow } from '../data/enemyTypes'
import type { PlayerSave } from '../save/types'
import { recordEnemyKill } from '../achievements/progress'
import { bossProfile, isBossAddFight, type BossProfile } from './boss'
import { getEnemy } from './engine'

export interface SquidlingVictoryResult {
  save: PlayerSave
  xpGained: number
  /** Always Fishing for Squidling kills. */
  xpSkillId: string
  message: string
  /** True when the boss resumes; false when another add remains. */
  bossResumed: boolean
}

export function beginBossAddsEncounter(
  db: GameDatabase,
  save: PlayerSave,
  bossEnemy: EnemyRow,
  profile: BossProfile,
  pendingHp: number,
  roundEndIso: string,
): PlayerSave {
  const squidling = getEnemy(db, profile.squidlingEnemyId!)
  if (!squidling) return save
  return {
    ...save,
    combatBossPendingId: bossEnemy['Enemy ID'],
    combatBossPendingHp: pendingHp,
    combatBossAddsRemaining: profile.squidlingCount,
    combatBossAddsTriggered: true,
    combatBossInkActive: false,
    combatEnemyId: squidling['Enemy ID'],
    combatEnemyHp: squidling['Maximum HP'],
    combatRoundStartedAt: roundEndIso,
    combatSkipEnemyAttack: false,
    combatBossSleepRoundsRemaining: null,
  }
}

export function applySquidlingVictory(
  db: GameDatabase,
  save: PlayerSave,
  squidling: EnemyRow,
  roundEndIso: string,
): SquidlingVictoryResult {
  const xpAmount = Number(squidling['Combat XP'] ?? 0)
  let next = applyXp(save, db, FISHING_SKILL_ID, xpAmount).save
  const kills = Number(next.statistics.values.monsters_killed ?? 0) + 1
  next = {
    ...next,
    statistics: {
      values: {
        ...next.statistics.values,
        monsters_killed: kills,
      },
    },
  }
  next = recordEnemyKill(db, next, squidling['Enemy ID'])

  const remaining = Math.max(0, (next.combatBossAddsRemaining ?? 1) - 1)
  const bossId = next.combatBossPendingId
  const pendingHp = next.combatBossPendingHp ?? 0

  if (remaining > 0 && bossId) {
    const profile = bossProfile(getEnemy(db, bossId))
    const nextSquidling =
      profile?.squidlingEnemyId != null ? getEnemy(db, profile.squidlingEnemyId) : undefined
    if (!nextSquidling) {
      return { save: next, xpGained: xpAmount, xpSkillId: FISHING_SKILL_ID, message: 'Squidling defeated.', bossResumed: false }
    }
    const total = profile?.squidlingCount ?? remaining + 1
    const defeated = total - remaining
    return {
      save: {
        ...next,
        combatBossAddsRemaining: remaining,
        combatEnemyId: nextSquidling['Enemy ID'],
        combatEnemyHp: nextSquidling['Maximum HP'],
        combatRoundStartedAt: roundEndIso,
        combatBossInkActive: false,
        combatSkipEnemyAttack: false,
      },
      xpGained: xpAmount,
      xpSkillId: FISHING_SKILL_ID,
      message: `Squidling defeated (${defeated}/${total}). Another emerges!`,
      bossResumed: false,
    }
  }

  if (!bossId) {
    return { save: next, xpGained: xpAmount, xpSkillId: FISHING_SKILL_ID, message: 'Squidling defeated.', bossResumed: false }
  }

  const boss = getEnemy(db, bossId)
  if (!boss) {
    return { save: next, xpGained: xpAmount, xpSkillId: FISHING_SKILL_ID, message: 'Squidling defeated.', bossResumed: false }
  }

  return {
    save: {
      ...next,
      combatBossAddsRemaining: null,
      combatBossPendingId: null,
      combatBossPendingHp: null,
      combatEnemyId: bossId,
      combatEnemyHp: pendingHp,
      combatRoundStartedAt: roundEndIso,
      combatBossInkActive: false,
      combatSkipEnemyAttack: false,
      combatBossSleepRoundsRemaining: null,
    },
    xpGained: xpAmount,
      xpSkillId: FISHING_SKILL_ID,
    message: `Squidling defeated. ${boss['Display Name']} returns at ${pendingHp} HP!`,
    bossResumed: true,
  }
}

export function isSquidlingVictory(save: PlayerSave, enemy: EnemyRow): boolean {
  if (!isBossAddFight(save)) return false
  return save.combatEnemyId === enemy['Enemy ID']
}
